File: Examples/AppNet/Example/ViewController.m
#import "ViewController.h"

#import "User.h"
#import "Data.h"

#import "DATASource.h"
#import "DATAStack.h"
#import "AppDelegate.h"
#import "Networking.h"

static NSString * const CellIdentifier = @"Cell";

@interface ViewController ()

@property (nonatomic, strong) DATASource *dataSource;
@property (nonatomic, strong) NSFetchedResultsController *fetchedResultsController;
@property (nonatomic, strong) DATAStack *dataStack;

@end

@implementation ViewController

- (instancetype)initWithDataStack:(DATAStack *)dataStack
{
    self = [super init];
    if (!self) return nil;

    _dataStack = dataStack;

    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:CellIdentifier];

    return self;
}

#pragma mark - Getters

- (DATASource *)dataSource
{
    if (_dataSource) return _dataSource;

    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"Data"];
    request.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"createdAt" ascending:YES]];

    _dataSource = [[DATASource alloc] initWithTableView:self.tableView
                                           fetchRequest:request
                                         cellIdentifier:CellIdentifier
                                            mainContext:self.dataStack.mainContext];

    _dataSource.configureCellBlock = ^(UITableViewCell *cell, Data *item, NSIndexPath *indexPath) {
        cell.textLabel.text = item.text;
    };

    return _dataSource;
}

#pragma mark - View Lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.tableView.dataSource = self.dataSource;
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];

    [self fetchData];
}

#pragma mark - Sync

- (void)fetchData
{
    Networking *networking = [[Networking alloc] initWithDataStack:self.dataStack];
    [networking fetchPosts];
}

@end

File: Examples/AppNet-Swift/AppNet-Swift/ViewController.swift
import UIKit

class ViewController: UITableViewController, UITableViewDelegate, UITableViewDataSource {
  let SYNCCellIdentifier = "CellID"
  let SYNCReloadTableNotification = "SYNCReloadTableNotification"

  let dataStack: DATAStack!
  var networking: Networking!
  var items = [Data]()

  // MARK: Initializers

  required init(dataStack: DATAStack) {
    super.init(nibName: nil, bundle: nil);

    self.dataStack = dataStack
  }

  required init(coder aDecoder: NSCoder) {
    super.init(nibName: nil, bundle: nil);
  }

  // MARK: View Lifecycle

  override func viewDidLoad() {
    super.viewDidLoad()

    self.title = "AppNet"
    self.tableView.registerClass(UITableViewCell.classForCoder(), forCellReuseIdentifier: SYNCCellIdentifier)
    self.tableView.allowsSelection = false

    self.items = Array()
    fetchCurrentObjects()
  }

  override func viewDidAppear(animated: Bool) {
    super.viewDidAppear(true)

    fetchNewData()
  }

  // MARK: TableView methods

  override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return items.count
  }

  override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
    var cell = self.tableView.dequeueReusableCellWithIdentifier(SYNCCellIdentifier) as UITableViewCell
    let data = self.items[indexPath.row]

    cell = UITableViewCell(style: UITableViewCellStyle.Subtitle, reuseIdentifier: SYNCCellIdentifier)

    cell.textLabel?.text = data.text
    cell.textLabel?.numberOfLines = 1000
    cell.detailTextLabel?.text = data.user.username

    return cell
  }

  // MARK: Networking methods

  func fetchNewData() {
    self.networking = Networking(dataStack: self.dataStack)
    self.networking.fetchNewContent { () -> Void in
      self.fetchCurrentObjects()
    }
  }

  // MARK: Model methods

  func fetchCurrentObjects() {
    let request = NSFetchRequest(entityName: "Data")
    request.sortDescriptors = [NSSortDescriptor(key: "createdAt", ascending: true)]
    self.items = self.dataStack.mainContext.executeFetchRequest(request, error: nil) as Array

    tableView.reloadData()
  }
}
